sdk/src/main/java/io/opentelemetry/sdk/tags/TagMapSdk.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.tags;
+
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.Tag;
+import io.opentelemetry.tags.TagKey;
+import io.opentelemetry.tags.TagMap;
+import io.opentelemetry.tags.TagMetadata;
+import io.opentelemetry.tags.TagValue;
+import io.opentelemetry.tags.unsafe.ContextUtils;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+public class TagMapSdk implements TagMap {
+
+  // The types of the TagKey and Tag must match for each entry.
+  private final Map<TagKey, Tag> tags;
+  @Nullable private final TagMap parent;
+
+  /**
+   * Creates a new {@link TagMapSdk} with the given tags.
+   *
+   * @param tags the initial tags for this {@code TagMapSdk}.
+   * @param parent providing a default set of tags
+   */
+  private TagMapSdk(Map<? extends TagKey, ? extends Tag> tags, TagMap parent) {
+    this.tags = Collections.unmodifiableMap(new HashMap<>(checkNotNull(tags, "tags")));
+    this.parent = parent;
+  }
+
+  @Override
+  public Iterator<Tag> getIterator() {
+    Map<TagKey, Tag> combined = new HashMap<>(tags);
+    if (parent != null) {
+      for (Iterator<Tag> it = parent.getIterator(); it.hasNext(); ) {
+        Tag tag = it.next();
+        if (!combined.containsKey(tag.getKey())) {
+          combined.put(tag.getKey(), tag);
+        }
+      }
+    }
+    // Clean out any null values that may have been added by Builder.remove.
+    for (Iterator<Tag> it = combined.values().iterator(); it.hasNext(); ) {
+      if (it.next() == null) {
+        it.remove();
+      }
+    }
+
+    return Collections.unmodifiableCollection(combined.values()).iterator();
```

Comment:
Is there any reason not to cache this value? The merged `Map`, that is. I wouldn't like to have a pair of passes/created values per calls to `getIterator()` ;)

Reply:
Can you propose a solution to cache this? Based on my understanding you have to always create a new iterator,

Reply:
I can cook a small PR later on, and we discuss its validity by then.